Benchmarks for Speedy/DAML-LF/Protobuf translations [DPP-104]

stefanobaghino-da opened the pull request:
A set of JMH micro-benchmarks that take a ledger export and measure how much time it takes to:

* decode and encode to Protobuf
* decode and encode to the DAML-LF representation
* decode and encode to the Speedy representation
* for both individual DAML-LF values and transactions

By default the benchmarks run on the reference ledger export. A parameter can be passed to point at a specific ledger export.

Followed the _boy scout rule_ here and there.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

remyhaemmerle-da commented on ledger/participant-state/kvutils/tools/BUILD.bazel:
```diff
@@ -125,3 +126,31 @@ da_scala_benchmark_jmh(
+    srcs = glob(["codec-benchmark/src/benchmark/scala/**/*.scala"]),
+    data = [
+        "//ledger/participant-state/kvutils:reference-ledger-export.out",
+    ] if not is_windows else [],
```
Should you not exclude the full rule in windows ?

remyhaemmerle-da commented on daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:
```diff
@@ -831,6 +831,20 @@ private[lf] object Speedy {
+    @throws[PackageNotFound]
+    @throws[CompilationError]
+    // Construct a machine for testing.
+    private[speedy] def dummy(
+        compiledPackages: CompiledPackages,
+    ): Machine = Machine(
+      compiledPackages = compiledPackages,
+      submissionTime = Time.Timestamp.MinValue,
+      initialSeeding = InitialSeeding.NoSeed,
+      expr = null,
+      globalCids = Set.empty,
+      committers = Set.empty,
```
is it really worthwhile to create a new builder if you use it only once ?

stefanobaghino-da replied:
I can remove it, I was hoping it was useful.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/8272/commits/bda089345a2d7c6d7d9dbac52a2374a6e5d2390a

remyhaemmerle-da commented:
Thanks for the typo fixes.

tudor-da commented on ledger/participant-state/kvutils/tools/BUILD.bazel:
```diff
@@ -125,3 +126,31 @@ da_scala_benchmark_jmh(
+        "//daml-lf/transaction:value_proto_java",
+        "//ledger/participant-state/kvutils",
+        "@maven//:com_google_protobuf_protobuf_java",
+        "@maven//:org_scalaz_scalaz_core_2_12",
```
Is scalaz used in this module?

stefanobaghino-da replied:
Good catch. No, and apparently Buildozer is disabled on `scala_benchmark_jmh` rule instances. Don't know why. Removing.

stefanobaghino-da replied:
Addressed by https://github.com/digital-asset/daml/pull/8272/commits/642cce9316172827f62a41ea6652f86d452a278a

stefanobaghino-da replied:
I'm an idiot, I removed the dependency from the wrong rule.

stefanobaghino-da replied:
Yes, it's needed. Reverting.

```
> bazebuildun //ledger/participant-state/kvutils/tools:benchmark-codec
INFO: Invocation ID: 1b743167-678b-4e7e-b09e-0e9a12ca65a3
INFO: Analyzed target //ledger/participant-state/kvutils/tools:benchmark-codec (1 packages loaded, 24 targets configured).
INFO: Found 1 target...
ERROR: /home/ste/Projects/digital-asset/daml/ledger/participant-state/kvutils/tools/BUILD.bazel:130:23: scala //ledger/participant-state/kvutils/tools:benchmark-codec_generator failed (Exit 1): scalac failed: error executing command 
  (cd /home/ste/.cache/bazel/_bazel_ste/74ca99b07c4389a5a21de1bd00106db4/execroot/com_github_digital_asset_daml && \
  exec env - \
  bazel-out/k8-opt/bin/external/io_bazel_rules_scala/src/java/io/bazel/rulesscala/scalac/scalac @bazel-out/k8-opt/bin/ledger/participant-state/kvutils/tools/benchmark-codec_generator_scalac_worker_input)
Execution platform: @io_tweag_rules_nixpkgs//nixpkgs/platforms:host
ledger/participant-state/kvutils/tools/codec-benchmark/src/benchmark/scala/com/daml/lf/DecodeTransactionBenchmark.scala:26: error: Symbol 'type scalaz.Equal' is missing from the classpath.
This symbol is required by 'method com.daml.lf.value.Value.VersionedValue.VersionedValue$u0020Equal$u0020instance'.
Make sure that type Equal is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
A full rebuild may help if 'Value.class' was compiled against an incompatible version of scalaz.
    encodedTransactions.map(assertDecode)
                            ^
one error found
one error found
Build failed
java.lang.RuntimeException: Build failed
	at io.bazel.rulesscala.scalac.ScalacWorker.compileScalaSources(ScalacWorker.java:259)
	at io.bazel.rulesscala.scalac.ScalacWorker.work(ScalacWorker.java:79)
	at io.bazel.rulesscala.worker.Worker.persistentWorkerMain(Worker.java:92)
	at io.bazel.rulesscala.worker.Worker.workerMain(Worker.java:46)
	at io.bazel.rulesscala.scalac.ScalacWorker.main(ScalacWorker.java:46)
Target //ledger/participant-state/kvutils/tools:benchmark-codec failed to build
INFO: Elapsed time: 1.517s, Critical Path: 1.03s
INFO: 0 processes.
FAILED: Build did NOT complete successfully
```

tudor-da replied:
Hmm. Ok, so Bazel modules don't see transitive deps. Didn't know that

stefanobaghino-da commented:
/azp run

stefanobaghino-da commented:
I have no idea of why this fails while building now, I simply removed the broken commit and re-run on a commit that already gave a green light before. 🤷🏻‍♂️

stefanobaghino-da commented:
> I have no idea of why this fails while building now, I simply removed the broken commit and re-run on a commit that already gave a green light before. 🤷🏻‍♂️

It actually didn't pass, but locally it builds. Seems like there's a missing dependency but I have no idea of why the same rule succeeds locally.

stefanobaghino-da commented:
Looks like the reason of the failure is that https://github.com/digital-asset/daml/pull/8273 removed `ValueVersions`, having a look at what I should use instead (looks like it's just `ValueVersion` instead of `ValueVersions`).

stefanobaghino-da commented:
> Looks like the reason of the failure is that #8273 removed `ValueVersions`, having a look at what I should use instead (looks like it's just `ValueVersion` instead of `ValueVersions`).

Worked, rebased on the latest main branch and fixed in https://github.com/digital-asset/daml/pull/8272/commits/c9bcdb43dcc396ee99a045eb35f398649eea93cd
